File: services/auth_service.go
package services

import (
	"clustta/internal/auth_service"
	"clustta/internal/error_service"
)

type AuthService struct{}

// AuthMode type alias for frontend binding
type AuthMode = auth_service.AuthMode

// Auth mode constants
const (
	AuthModeGlobal  = auth_service.AuthModeGlobal
	AuthModeStudio  = auth_service.AuthModeStudio
	AuthModeOffline = auth_service.AuthModeOffline
)

// Login authenticates a user with username and password against Clustta Cloud.
// Returns the authentication token or an error if login fails.
func (a *AuthService) Login(username, password string) (auth_service.Token, error) {
	token, err := auth_service.Login(username, password)
	if err != nil {
		return token, err
	}
	return token, nil
}

// LoginWithHost authenticates a user against a specified authentication host.
// authMode should be "global", "studio", or "offline".
// Returns the authentication token or an error if login fails.
func (a *AuthService) LoginWithHost(username, password, authHost string, authMode string, studioId string) (auth_service.Token, error) {
	mode := auth_service.AuthMode(authMode)
	if !auth_service.IsValidAuthMode(mode) {
		mode = auth_service.AuthModeGlobal
	}
	token, err := auth_service.LoginWithHost(username, password, authHost, mode, studioId)
	if err != nil {
		return token, err
	}
	return token, nil
}

// EnableOfflineMode sets up offline mode without authentication.
// Creates a local-only pseudo-account.
func (a *AuthService) EnableOfflineMode() error {
	return auth_service.EnableOfflineMode()
}

// IsOfflineMode checks if the current session is in offline mode.
func (a *AuthService) IsOfflineMode() bool {
	return auth_service.IsOfflineMode()
}

// GetAuthMode returns the current authentication mode.
// Returns "global", "studio", or "offline".
func (a *AuthService) GetAuthMode() string {
	return string(auth_service.GetActiveAuthMode())
}

// GetAuthHost returns the current authentication host URL.
// Returns empty string if in offline mode.
func (a *AuthService) GetAuthHost() string {
	return auth_service.GetAuthHost()
}

// Register creates a new user account on Clustta Cloud.
// Returns the created user or an error if registration fails.
func (a *AuthService) Register(firstName, lastName, username, email, password, confirmPassword string) (auth_service.User, error) {
	user, err := auth_service.Register(firstName, lastName, username, email, password, confirmPassword)
	if err != nil {
		return user, err
	}
	return user, nil
}

// RegisterWithHost creates a new user account on a specified authentication host.
// Returns the created user or an error if registration fails.
func (a *AuthService) RegisterWithHost(firstName, lastName, username, email, password, confirmPassword, authHost string) (auth_service.User, error) {
	user, err := auth_service.RegisterWithHost(firstName, lastName, username, email, password, confirmPassword, authHost)
	if err != nil {
		return user, err
	}
	return user, nil
}

// UpdateUser updates the current user's profile information.
// Returns the updated user or an error if the update fails.
func (a *AuthService) UpdateUser(firstName, lastName, username, email string) (auth_service.User, error) {
	user, err := auth_service.UpdateUser(firstName, lastName, username, email)
	if err != nil {
		return user, err
	}
	return user, nil
}

// UpdateUserPhoto updates the current user's profile photo.
// Returns an error if the upload fails.
func (a *AuthService) UpdateUserPhoto(photo []byte) error {
	return auth_service.UpdateUserPhoto(photo)
}

// Logout ends the current user session.
// Returns an error if logout fails.
func (a *AuthService) Logout(username, password string) error {
	err := auth_service.Logout()
	if err != nil {
		return err
	}
	return nil
}

// AuthUser retrieves the currently authenticated user.
// Returns an error if no user is found or the token is invalid.
func (a *AuthService) AuthUser() (auth_service.User, error) {
	saveToken, err := auth_service.GetToken()
	if err != nil {
		return auth_service.User{}, err
	}
	user := saveToken.User
	if user.Username == "" {
		return user, error_service.ErrUserNotFound
	}
	return user, nil
}

// IsAuthenticated checks if a user is currently authenticated.
// Returns authentication status, user data, and any error encountered.
func (a *AuthService) IsAuthenticated() (bool, auth_service.User, error) {
	isAuthenticated, err := auth_service.IsAuthenticated()
	if err != nil {
		if err.Error() == "secret not found in keyring" {
			return false, auth_service.User{}, nil
		}
		return false, auth_service.User{}, err
	}
	if isAuthenticated {
		saveToken, err := auth_service.GetToken()
		if err != nil {
			if err.Error() == "secret not found in keyring" {
				return false, auth_service.User{}, nil
			}
			return false, auth_service.User{}, err
		}
		user := saveToken.User
		if user.Username == "" {
			return false, auth_service.User{}, nil
		}
		return true, user, nil
	}
	return false, auth_service.User{}, nil
}

// CheckUsernameExists checks if a username is already registered.
// Returns true if the username exists, false otherwise.
func (a *AuthService) CheckUsernameExists(username string) (bool, error) {
	return auth_service.CheckUsernameExists(username)
}

// CheckEmailExists checks if an email address is already registered.
// Returns true if the email exists, false otherwise.
func (a *AuthService) CheckEmailExists(email string) (bool, error) {
	return auth_service.CheckEmailExists(email)
}

// DeactivateUserAccount deactivates the current user's account.
// Returns an error if the deactivation fails.
func (a *AuthService) DeactivateUserAccount() error {
	return auth_service.DeactivateUserAccount()
}

// SendInvitationEmail sends a project invitation to an email address.
// Returns an error if the email send fails.
func (a *AuthService) SendInvitationEmail(email, studioName, projectName string) error {
	return auth_service.SendInvitationEmail(email, studioName, projectName)
}

// VerifyOTP verifies a one-time password token.
// Returns an error if verification fails.
func (a *AuthService) VerifyOTP(email, token string) error {
	return auth_service.VerifyOTP(email, token)
}

// ResendToken resends the verification token to an email address.
// Returns an error if the send fails.
func (a *AuthService) ResendToken(email string) error {
	return auth_service.ResendToken(email)
}

// ChangePassword changes the current user's password.
// Returns an error if the password change fails.
func (a *AuthService) ChangePassword(currentPassword, newPassword, confirmPassword string) error {
	return auth_service.ChangePassword(currentPassword, newPassword, confirmPassword)
}

// ResetPassword sends a password reset email to the specified email address.
// Returns an error if the reset request fails.
func (a *AuthService) ResetPassword(email string) error {
	return auth_service.ResetPassword(email)
}

File: frontend/src/services/adapters/authservice.js
import { STORAGE_KEYS } from './config.js';
import { globalApiCall } from './http-client.js';
import {
  addAccountToStorage,
  removeAccountFromStorage,
  getMultiAccountToken,
  clearUserSpecificData,
  clearAllUserData,
} from './storage.js';

export const AuthService = {
  // Authenticates user with email and password
  Login: async (username, password) => {
    const response = await globalApiCall('/auth/login', 'POST', { email: username, password });

    const user = {
      id: response.user?.id || '',
      username: response.user?.username || response.user?.user_name || '',
      email: response.user?.email || '',
      first_name: response.user?.first_name || '',
      last_name: response.user?.last_name || '',
      photo: response.user?.photo || '',
    };

    const token = {
      session_id: response.session_id || '',
      user: user,
    };

    if (response.session_id) {
      localStorage.setItem(STORAGE_KEYS.SESSION_ID, response.session_id);
    }
    localStorage.setItem(STORAGE_KEYS.USER, JSON.stringify(user));
    addAccountToStorage(token);

    return token;
  },

  // Logs out the current user and switches to next account if available
  Logout: async () => {
    try {
      await globalApiCall('/auth/logout', 'GET');
    } catch (error) {
      console.warn('Logout API call failed:', error);
    }

    const multiToken = getMultiAccountToken();
    const activeAccountId = multiToken.active_account_id;

    if (activeAccountId) {
      removeAccountFromStorage(activeAccountId);
    }

    const updatedMultiToken = getMultiAccountToken();
    const remainingAccounts = Object.keys(updatedMultiToken.accounts);

    if (remainingAccounts.length > 0) {
      const nextAccountId = remainingAccounts[0];
      const nextAccount = updatedMultiToken.accounts[nextAccountId];

      localStorage.setItem(STORAGE_KEYS.USER, JSON.stringify(nextAccount.user));
      if (nextAccount.session_id) {
        localStorage.setItem(STORAGE_KEYS.SESSION_ID, nextAccount.session_id);
      }

      clearUserSpecificData();
      return { hasRemainingAccounts: true, nextAccount: nextAccount };
    } else {
      clearAllUserData();
      return { hasRemainingAccounts: false, nextAccount: null };
    }
  },

  // Logs out a specific account without affecting current session
  LogoutAccount: async (userId) => {
    const multiToken = getMultiAccountToken();

    if (multiToken.active_account_id === userId) {
      return await AuthService.Logout();
    }

    removeAccountFromStorage(userId);
    return { hasRemainingAccounts: Object.keys(getMultiAccountToken().accounts).length > 0 };
  },

  // Registers a new user account
  Register: async (firstName, lastName, username, email, password, confirmPassword) => {
    const response = await globalApiCall('/auth/register', 'POST', {
      first_name: firstName,
      last_name: lastName,
      username,
      email,
      password,
      confirm_password: confirmPassword,
    });
    return {
      id: response.id || '',
      username: response.username || response.user_name || '',
      email: response.email || '',
      first_name: response.first_name || '',
      last_name: response.last_name || '',
      photo: '',
    };
  },

  // Checks if current session is authenticated
  IsAuthenticated: async () => {
    try {
      const localUser = JSON.parse(localStorage.getItem(STORAGE_KEYS.USER) || 'null');

      if (!localUser || !localUser.id) {
        return [false, {}];
      }

      await globalApiCall('/auth/authenticated', 'GET');
      return [true, localUser];
    } catch (error) {
      return [false, {}];
    }
  },

  // Returns the authenticated user or throws if not authenticated
  AuthUser: async () => {
    const localUser = JSON.parse(localStorage.getItem(STORAGE_KEYS.USER) || 'null');

    if (!localUser || !localUser.id) {
      throw new Error('Not authenticated');
    }

    try {
      await globalApiCall('/auth/authenticated', 'GET');
      return localUser;
    } catch (error) {
      clearAllUserData();
      throw new Error('Session expired');
    }
  },

  // Verifies OTP token for email verification
  VerifyOTP: async (email, token) => {
    await globalApiCall('/auth/verify-otp', 'POST', { email, otp: token });
  },

  // Resends verification token to email
  ResendToken: async (email) => {
    await globalApiCall('/auth/token/resend', 'POST', { email });
  },

  // Sends password reset email
  ResetPassword: async (email) => {
    await globalApiCall('/auth/reset-password', 'POST', { email });
  },

  // Changes password using reset token (from email link)
  ResetChangePassword: async (email, token, newPassword, confirmPassword) => {
    await globalApiCall('/auth/reset-password', 'PUT', {
      email,
      token,
      new_password: newPassword,
      confirm_password: confirmPassword,
    });
  },

  // Changes user password
  ChangePassword: async (currentPassword, newPassword, confirmPassword) => {
    await globalApiCall('/auth/change-password', 'POST', {
      password: currentPassword,
      new_password: newPassword,
      confirm_password: confirmPassword,
    });
  },

  // Checks if email is already registered
  CheckEmailExists: async (email) => {
    try {
      const response = await globalApiCall(`/auth/email-exists/${encodeURIComponent(email)}`, 'GET');
      return response.exists === true;
    } catch {
      return false;
    }
  },

  // Checks if username is already taken
  CheckUsernameExists: async (username) => {
    try {
      const response = await globalApiCall(`/auth/username-exists/${encodeURIComponent(username)}`, 'GET');
      return response.exists === true;
    } catch {
      return false;
    }
  },

  // Updates user profile information
  UpdateUser: async (firstName, lastName, username, email) => {
    const response = await globalApiCall('/person/update', 'PUT', {
      first_name: firstName,
      last_name: lastName,
      username,
      email,
    });
    const user = {
      id: response.id || '',
      username: response.username || response.user_name || '',
      email: response.email || '',
      first_name: response.first_name || '',
      last_name: response.last_name || '',
      photo: response.photo || '',
    };
    localStorage.setItem(STORAGE_KEYS.USER, JSON.stringify(user));
    return user;
  },

  // Updates user profile photo
  UpdateUserPhoto: async (photoBase64) => {
    await globalApiCall('/person/photo', 'POST', { photo: photoBase64 });
  },

  // Deactivates the user account
  DeactivateUserAccount: async () => {
    await globalApiCall('/person/deactivate-account', 'POST');
    localStorage.removeItem(STORAGE_KEYS.SESSION_ID);
    localStorage.removeItem(STORAGE_KEYS.USER);
  },

  // Sends invitation email to collaborate on a project
  SendInvitationEmail: async (email, studioName, projectName) => {
    await globalApiCall('/auth/send-invitation', 'POST', { email, studio_name: studioName, project_name: projectName });
  },
};
